### PR TITLE
✨ Add ItemTapped action

### LINKED ISSUE #270: Add the ItemTapped action
## Description

Add the ItemTapped action and handle it in the reducer. This should simply set the selected item in the state.

## Definition of done
- [ ] The `selectedItem` is set in the state with an editable time entry
- [ ] Tests are created

### PR DESCRIPTION
**Pull request by semanticer**
### Summary
This PR implements ItemTapped action as described in #270 

### Technical considerations
Nothing really interesting

### Why is it done like this and not some other way?
`EditableTimeEntry` and `EditableProject` were moved to `com.toggl.models.domain` because I needed to use them in the Calendar Module. Let me know if you find a better place for these two orphans.

Instead of using `Either<TimeEntry, CalendarEvent>` for selectedItem, I created dedicated sealed type `SelectedCalendarItem`

### Relationships

Closes #270

### Pull Request checklist

- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [x] One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR

### Pointers
It's fairly simple PR but going commit by commit will make the reviewing more pleasant.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
🦑 it

### CODE REVIEW

File: calendar/src/main/java/com/toggl/calendar/domain/CalendarState.kt [hunk 1]
```diff
@@ -1,10 +1,12 @@
 package com.toggl.calendar.domain
 
 import arrow.optics.optics
+import com.toggl.models.domain.Task
 
 @optics
 data class CalendarState(
-    val toBeDeleted: String
+    val tasks: Map<Long, Task>,
```
Comment (juanlaube) [inline calendar/src/main/java/com/toggl/calendar/domain/CalendarState.kt:8]: Why do we need tasks here? 🤔

Comment (semanticer) [reply calendar/src/main/java/com/toggl/calendar/domain/CalendarState.kt]: I was probably in a very wierd state of mind when I wrote this 🤔

File: timer/src/main/java/com/toggl/timer/startedit/domain/ProjectTagChipSelector.kt [hunk 2]
```diff
@@ -15,11 +15,11 @@ fun projectTagChipSelector(
...
-        val project = projects[editableTimeEntry.projectId] ?: throw ProjectDoesNotExistException()
-        yield(ChipViewModel.Project(project))
-    } else {
-        yield(ChipViewModel.AddProject(addProjectLabel))
+    editableTimeEntry.projectId.let { projectId ->
+        if (projectId != null) {
+            val project = projects[projectId] ?: throw ProjectDoesNotExistException()
+            yield(ChipViewModel.Project(project))
+        } else yield(ChipViewModel.AddProject(addProjectLabel))
     }
```
Comment (daividssilverio) [inline timer/src/main/java/com/toggl/timer/startedit/domain/ProjectTagChipSelector.kt:23]:
```suggestion
    editableTimeEntry.projectId?.let { projectId ->
        val project = projects[projectId] ?: throw ProjectDoesNotExistException()
        yield(ChipViewModel.Project(project))
    } ?: yield(ChipViewModel.AddProject(addProjectLabel))
```

The elvis operator looks alright for this case. Doesn't provide much of a readability improvement though.

File: timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt [hunk 3]
```diff
@@ -318,7 +318,11 @@ class StartEditReducer @Inject constructor(
 
     private fun StartEditState.modifyWithCreateProjectSuggestion(autocompleteSuggestion: AutocompleteSuggestion.CreateProject): StartEditState =
         StartEditState.editableTimeEntry.modify(this) {
-            it.copy(editableProject = EditableProject(name = autocompleteSuggestion.name, workspaceId = 1))
+            it.copy(editableProject = EditableProject(
+                name = autocompleteSuggestion.name,
+                workspaceId = 1
+            )
+            )
```
Comment (daividssilverio) [inline timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt:325]: 👀

File: timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt [hunk 4]
```diff
@@ -462,11 +466,13 @@ class StartEditReducer @Inject constructor(
...
-
-        val relativeDuration = duration ?: startTime.absoluteDurationBetween(now)
-        return startTime + relativeDuration
+        return startTime.let { startTime ->
+            if (startTime == null) now
+            else {
+                val relativeDuration = duration ?: startTime.absoluteDurationBetween(now)
+                startTime + relativeDuration
+            }
+        }
```
Comment (daividssilverio) [inline timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt:475]:
```suggestion
        return startTime?.let { startTime ->
                val relativeDuration = duration ?: startTime.absoluteDurationBetween(now)
                startTime + relativeDuration
            } ?: now
        }
```

Another opportunity 🤔 What is our standing on the elvis operator again? 🤔 Maybe we should discuss when we should use it 🤔

Comment (semanticer) [reply timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt]: I'm not against, especially when it helps reduce nesting.

File: timer/src/main/java/com/toggl/timer/startedit/domain/ProjectTagChipSelector.kt [hunk 5]
```diff
@@ -15,17 +15,15 @@ fun projectTagChipSelector(
...
 
-    if (editableTimeEntry.projectId != null) {
-        val project = projects[editableTimeEntry.projectId] ?: throw ProjectDoesNotExistException()
+    editableTimeEntry.projectId?.let { projectId ->
+        val project = projects[projectId] ?: throw ProjectDoesNotExistException()
         yield(ChipViewModel.Project(project))
-    } else {
-        yield(ChipViewModel.AddProject(addProjectLabel))
-    }
+    } ?: yield(ChipViewModel.AddProject(addProjectLabel))
```
Comment (heytherewill) [inline timer/src/main/java/com/toggl/timer/startedit/domain/ProjectTagChipSelector.kt:21]: This change is completely unrelated to this PR. Also, using a null-coalescence PR for a side effect (yielding) reads really odd

Comment (semanticer) [reply timer/src/main/java/com/toggl/timer/startedit/domain/ProjectTagChipSelector.kt:21]: It is related. `EditableTimeEntry` and `EditableProject` were moved to a different module and that broke the smart cast magic here. 
Side effect yield is indeed suboptimal but my original alternative was maybe even worse 🙈
